Feature/last used row column

ColmBhandal opened the pull request:
## Description

Added functions to 2D arrays allowing for the last used row/column to be calculated.

ColmBhandal committed:
Skeletal functions and failing tests

ColmBhandal committed:
Implemented the general purpose functions in terms of fold and find last index

The fold and find last index functions already existed
Fold is part of this library
Find last index is a static function on the Array class
The fold we do chooses whichever element satisfies the predicate - if neither does then the choice doesn't matter

ColmBhandal committed:
Change log update

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -1,11 +1,33 @@
+        }
+        public static int LastUsedColumn<TVal>(this TVal[,] array, Predicate<TVal> isUsed)
+        {
+            TVal[] folded = array.FoldToSingleRow((v1, v2) => isUsed(v1) ? v1 : v2);
```
Would it not be more efficient (and maybe clearer) to simply iterate over the entire 2D array backwards?

The way you have it implemented now you are always iterating over the entire 2D array for the fold (plus running the `isUsed` function), even if the last column is used.

hctrdev commented on CsharpExtras/Extensions/ArrayExtension2D.cs:
```diff
@@ -1,11 +1,33 @@
+        {
+            return array.LastUsedColumn(s => !string.IsNullOrWhiteSpace(s));
+        }
+        public static int LastUsedColumn<TVal>(this TVal[,] array, Predicate<TVal> isUsed)
```
I would add some documentation on these new methods. I'm assuming they return the index of the last column where all the values match the function, is that right? What do you return if none match?

ColmBhandal committed:
Replaced usage of fold with more efficient functions and improved documentation


New functions return the last row/column where a given predicate over the entire row/column is true
Using the new functions, the usage predicate for a given row/column is whether there exists a single element in that row/column that's used
Added documentation to make it clear that the usage predicate is existentially, rather than universally, quantified to build the row/column predicate
